Extract additional information from sideloaded media

stevegrunwell opened the pull request:
Newer versions of Airstory (currently in beta) make it easier to set attributes (alt, title, etc.) on media within Airstory, which we want to capture when we sideload the media. As a result, image parsing has gone from simple string replacement to DOMDocument-driven parsing and attribute replacement.

New attributes we're tracking:
* post_author (based on the user importing the attachment)
* alt text

stevegrunwell committed:
Rewrite Formatting\sideload_images() to use DOMDocument instead of regular expressions

stevegrunwell committed:
Explicitly require the dom PHP extension for FormattingTest

stevegrunwell committed:
When sideloading media, attempt to set the attachment's post_author based on the parent post's author. Fixes #13

stevegrunwell committed:
Add Formatting\get_attachment_id_by_url(), which lets us retrieve an attachment ID based on its GUID. Based on https://pippinsplugins.com/retrieve-attachment-id-from-image-url/

stevegrunwell committed:
Merge branch 'develop' into feature/extract-media-information-when-sideloading

stevegrunwell committed:
When an image is sideloaded, recover the ID of the newly-created attachment, then store the alt attribute and original Airstory URL in post meta. Fixes #16

stevegrunwell commented:
@norcross This should be the last PR for the day, and it's largely switching from regex string replacements to DOMDocument when sideloading media on import.

norcross commented on includes/formatting.php:
```diff
@@ -78,30 +78,83 @@ function sideload_images( $post_id ) {
 
-	foreach ( array_unique( $matches['1'] ) as $remote ) {
-		$local = media_sideload_image( esc_url_raw( $remote ), $post_id, null, 'src' );
+	foreach ( $images as $image ) {
```
needs a check to make sure images exist before going into the process

norcross commented on includes/formatting.php:
```diff
@@ -78,30 +78,83 @@ function sideload_images( $post_id ) {
+
+		} else {
+			// Sideload the media.
+			$local = media_sideload_image( esc_url_raw( $src ), $post_id, null, 'src' );
```
if you use `media_handle_sideload` instead of `media_sideload_image`, you'll have the attachment ID which will avoid the need to use the `get_attachment_id_by_url` function.

the setup is a bit different, but here's a working example: https://github.com/norcross/external-image-upload/blob/master/lib/helper.php#L89

stevegrunwell replied:
Awesome, I'll rework it to use `media_handle_sideload()` instead, then — as you mentioned, GUIDs are flakey.

norcross replied:
side bonus: you can set the various parts of the image data (title, etc) using a function like I've done here: https://github.com/norcross/external-image-upload/blob/master/lib/helper.php#L255

norcross commented on includes/formatting.php:
```diff
@@ -122,3 +175,47 @@ function strip_wrapping_div( $content ) {
+	global $wpdb;
+
+	return (int) $wpdb->get_var( $wpdb->prepare(
+		"SELECT ID FROM $wpdb->posts WHERE post_type = 'attachment' AND guid = %s LIMIT 1;",
```
I'm sure you know, but this is kinda flaky, especially since the GUIDs can be changed or modified by other things.

stevegrunwell committed:
Instead of using `media_sideload_image()`, followed immediately by a lookup to determine the newly-inserted image's ID, use `media_handle_sideload()`, which returns the ID directly but requires a bit more setup.

For the sake of testability, the sideloading has been abstracted into Formatting\sideload_image(), which also accepts an array of meta key:value pairs to add via `update_post_meta()` — this is useful for alt text and any other attributes we might collect in the future.

stevegrunwell committed:
It's highly unlikely, but before we risk setting an image's src attribute to an empty string, ensure the value we get back from wp_get_attachment_url() is...well, not empty.

stevegrunwell committed:
Merge branch 'develop' into feature/extract-media-information-when-sideloading

norcross commented on includes/formatting.php:
```diff
@@ -57,6 +57,60 @@ function get_body_contents( $content ) {
+	 * Finally, store post meta. We'll always set _airstory_origin (the original image URL), but any
+	 * non-empty values in $metadata will also be set.
+	 */
+	add_post_meta( $image_id, '_airstory_origin', esc_url( $url ) );
```
why `add_post_meta` here, but `update_post_meta` a few lines later?

stevegrunwell replied:
I've encountered plugins before that have attempted to clean up image alt text, so it's possible that meta may already be set. The `_airstory_origin` meta key, however, is unique to this plugin and should be immutable. Using `add_post_meta()` just saves the check that `update_post_meta()` does to see if the key already exists.

norcross commented on includes/formatting.php:
```diff
@@ -57,6 +57,60 @@ function get_body_contents( $content ) {
+	require_once ABSPATH . 'wp-admin/includes/media.php';
+	require_once ABSPATH . 'wp-admin/includes/file.php';
+	require_once ABSPATH . 'wp-admin/includes/image.php';
+
```
should add a check here to make sure an URL was passed

stevegrunwell replied:
👍

norcross commented on includes/formatting.php:
```diff
@@ -57,6 +57,60 @@ function get_body_contents( $content ) {
+	 */
+	add_post_meta( $image_id, '_airstory_origin', esc_url( $url ) );
+
+	foreach ( (array) $metadata as $meta_key => $meta_value ) {
```
check for the `$metadata` to exist before starting the loop

stevegrunwell replied:
Casting it to an array will result in an empty array if it isn't already, which basically causes:

```php
foreach ( array() as $key => $value ) {
  // Nothing here will get run.
}
```

Not a major concern, IMO, but you're not wrong. I've added an `empty()` check around the block.

norcross commented on includes/formatting.php:
```diff
@@ -57,6 +57,60 @@ function get_body_contents( $content ) {
+ *                         has been created. These keys and values are assumed to be sanitized.
+ *                         Default is an empty array.
+ */
+function sideload_image( $url, $post_id = 0, $metadata = array() ) {
```
having `sideload_image` and `sideload_images` may get a bit confusing. perhaps `sideload_single_image` and `sideload_all_images` instead?

stevegrunwell replied:
Pssh, and make sense? Why would anyone do that? ;)

norcross commented on includes/formatting.php:
```diff
@@ -57,6 +57,60 @@ function get_body_contents( $content ) {
+function sideload_image( $url, $post_id = 0, $metadata = array() ) {
+	require_once ABSPATH . 'wp-admin/includes/media.php';
+	require_once ABSPATH . 'wp-admin/includes/file.php';
+	require_once ABSPATH . 'wp-admin/includes/image.php';
```
move these to below the `$tmp_file` check to avoid loading if not needed

stevegrunwell replied:
I had the same thought, but [`download_url()` is defined in wp-admin/includes/file.php](https://developer.wordpress.org/reference/functions/download_url/#source) — in the interest of not having multiple `require_once` blocks, I've kept them together at the top.

norcross replied:
I was not aware of that.

norcross reviewed (commented):
not all comments are blockers, but ready for update

stevegrunwell committed:
To avoid confusion, rename `sideload_images()` to `sideload_all_images()`, and `sideload_image()` to `sideload_single_image()`.

stevegrunwell committed:
Bail out super-early in Formatting\sideload_single_image() if the passed URL is not, in fact, a URL

stevegrunwell committed:
Rather than `foreach ( (array) $somethingThatMayOrMayNotBeEmpty as $k => $v ) { ... }`, wrap the loop in a quick `empty()` check.

stevegrunwell commented:
@norcross Changes have been addressed, appreciate you keeping me honest 💯
